MudExpansionPanel: Use ParameterState, remove obsolete API and change to async API

ScarletKuro opened the pull request:
## Description
Fixes: https://github.com/MudBlazor/MudBlazor/issues/8429

### Breaking Changes:
#### Remove `IsInitiallyExpanded`
It's not necessary anymore.
You can either set a constant
```razor
<MudExpansionPanels MultiExpansion>
    <MudExpansionPanel Class="panel-one" Text="Panel One" IsExpanded="true">
        ...
    </MudExpansionPanel>
</MudExpansionPanels>
```
`StateHasChanged` will not reset on re-render back to constant because we avoided writing to the Parameter https://learn.microsoft.com/en-us/aspnet/core/blazor/components/overwriting-parameters?view=aspnetcore-8.0

or you can use two way binding.
```razor
<MudExpansionPanels MultiExpansion>
    <MudExpansionPanel Class="panel-one" Text="Panel One" @bind-IsExpanded="_isExpanded">
        ...
    </MudExpansionPanel>
</MudExpansionPanels>

@code
{
    private bool _isExpanded = true;
}
```
To set the initial expansion.

#### Change API to be async.
This component was firing `EventCallback`s in a fire-and-forget manner. We changed all the API to be async.

#### Simplify code
We do not need `_collapseIsExpanded` variable or `updateParent` arguments, since the `ParamterState` manages the state correctly, therefore everything can be done in a more simple way. 

## How Has This Been Tested?
Visual, current unit tests, new unit test.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

ScarletKuro committed:
MudExpansionPanel: Use ParameterState, remove obsolete API and change to async API

henon commented:
I don't understand why we need the `IExpansionPanelManager` interface at all. Why not simply use internal methods like before? If I understand correctly, this change doesn't concern the users at all as the interface is internal anyway.

henon commented:
> Using only IsExpandedChanged is considered invalid now.

If you only care about getting notifications I think this should still be valid usage?

ScarletKuro commented:
> I don't understand why we need the `IExpansionPanelManager` interface at all. Why not simply use internal methods like before? If I understand correctly, this change doesn't concern the users at all as the interface is internal anyway.

True :joy: I just saw it was using `event` so I automatically put an interface because I got this observer-observable pattern pattern in my head, I will change that.

ScarletKuro committed:
Remove interface and use internal

ScarletKuro commented:
> If you only care about getting notifications I think this should still be valid usage?

It will not work currently if you just subscribe only to `IsExpandedChanged` and do not update the `IsExpanded`, it will not collapse or expand (depending on the initial `IsExpanded`). At first I thought everything is working as intended and it only worked because we modified `IsExpanded` directly and when i set it in the component this effect disappeared, but now after debugging I released it's ParamterState framework bug :triumph:. I know how to fix it, but I need to come up with integration tests since we didnt catch this bug before and it's not something you can unit tests as you need the re-renders etc.

ScarletKuro committed:
revert example as now it works fine

ScarletKuro commented:
Fixed: https://github.com/MudBlazor/MudBlazor/pull/8457
Rebased this branch, reverted the example with just `IsExpandedChanged`, and changed the PR description.
